File: Sources/structly/04_linked_list/LinkedListValues.swift
//
//  LinkedListValues.swift
//  MyCLI
//
//  Created by new on 3/15/26.
//

func linkedListValues2<T>(_ head: Node<T>?) -> [T] {
    var vals = [T]()
    var node = head
    while node != nil {
        vals.append(node!.val)
        node = node?.next
    }

    return vals
}

func linkedListValues<T>(_ head: Node<T>?) -> [T] {
    func rec(node: Node<T>?, acc: inout [T]) {
        guard let node else {
            return
        }
        acc.append(node.val)
        rec(node: node.next, acc: &acc)
    }
    var acc: [T] = []
    rec(node: head, acc: &acc)
    
    return acc
}

File: Sources/structlyC/04_linked_list/linked_list_values.cpp
//
//  linked_list_values.cpp
//  MyCLI
//
//  Created by new on 3/15/26.
//

#include <string>
#include <vector>

class Node {
public:
    std::string val;
    Node *next;

    Node(std::string initialVal) {
        val = initialVal;
        next = nullptr;
    }
};

std::vector<std::string> linkedListValuesIter(Node *head) {
    Node *node = head;
    std::vector<std::string> vals{};
    while (node != nullptr) {
        vals.push_back(node->val);
        node = node->next;
    }
    return vals;
}

void linked_list_values_r(Node *head, std::vector<std::string> &values) {
    if (head == nullptr) {
        return;
    }
    values.push_back(head->val);
    linked_list_values_r(head->next, values);
}

std::vector<std::string> linkedListValuesRec(Node *head) {
    std::vector<std::string> values{};
    linked_list_values_r(head, values);
    return values;
}
